pkg/criticalresource-admissionwebhook/webhook.go

```diff
@@ -0,0 +1,292 @@
+	providerGVR := schema.GroupVersionResource{
+		Group:    criticalResource.Spec.Provider.GroupResource.Group,
+		Version:  "v1",
+		Resource: criticalResource.Spec.Provider.GroupResource.Resource,
+	}
+	_, err := a.dynamicClient.Resource(providerGVR).Namespace(criticalResource.Namespace).Get(ctx, criticalResource.Spec.Provider.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	// TODO we can decide about finalized or not later
+	//if providerResource.GetDeletionTimestamp() != nil {
+	//	return nil
+	//}
+
+	return fmt.Errorf("%v named %q in namespace/%v still exists", providerGVR, criticalResource.Spec.Provider.Name, criticalResource.Namespace)
+}
+
+func (a *CriticalResourceAdmissionWebhook) validateAllCriteriaMet(ctx context.Context, criticalResource *apiserverv1.CriticalResource) []string {
+	messages := []string{}
+	for _, criteria := range criticalResource.Spec.Criteria {
+		if err := a.validateCriteriaMet(ctx, criticalResource.Namespace, criteria); err != nil {
+			messages = append(messages, err.Error())
+		}
+	}
+
+	return messages
+}
+
+func (a *CriticalResourceAdmissionWebhook) validateCriteriaMet(ctx context.Context, namespace string, criteria apiserverv1.CriticalResourceCriteria) error {
+	switch criteria.Type {
+	case apiserverv1.FinalizerType:
+		return a.validateFinalizerCriteriaMet(ctx, criteria)
+	case apiserverv1.SpecificResourceType:
+		return a.validateSpecificResourceCriteriaMet(ctx, namespace, criteria)
+	default:
+		return fmt.Errorf("%q is not a known type", criteria.Type)
+	}
+}
+
+func (a *CriticalResourceAdmissionWebhook) validateFinalizerCriteriaMet(ctx context.Context, criteria apiserverv1.CriticalResourceCriteria) error {
+	// TODO fix API to include version
+	gvr := schema.GroupVersionResource{
+		Group:    criteria.Finalizer.Group,
+		Version:  "v1",
+		Resource: criteria.Finalizer.Resource,
+	}
+	instanceList, err := a.dynamicClient.Resource(gvr).Namespace("" /*get instances in every namespace*/).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to list %v: %w", gvr, err)
+	}
+
+	numWithFinalizer := 0
+	for _, instance := range instanceList.Items {
+		for _, finalizer := range instance.GetFinalizers() {
+			if finalizer == criteria.Finalizer.FinalizerName {
+				numWithFinalizer++
+				break
+			}
+		}
+	}
+	if numWithFinalizer == 0 {
+		return nil
+	}
+	return fmt.Errorf("%d instances of %v contain finalizer", numWithFinalizer, gvr)
+}
+
+func (a *CriticalResourceAdmissionWebhook) validateSpecificResourceCriteriaMet(ctx context.Context, namespace string, criteria apiserverv1.CriticalResourceCriteria) error {
+	// TODO fix API to include version
+	gvr := schema.GroupVersionResource{
+		Group:    criteria.SpecificResource.Group,
+		Version:  "v1",
+		Resource: criteria.SpecificResource.Resource,
+	}
+	_, err := a.dynamicClient.Resource(gvr).Namespace(namespace).Get(ctx, criteria.SpecificResource.Name, metav1.GetOptions{})
```

Comment:
i think we have to check if the related resource is namespace or cluster scoped. I tested with
```
apiVersion: apiserver.openshift.io/v1
kind: CriticalResource
metadata:
  name: protect-deployment
spec:
  provider:
    group: apps
    resource: deployments
    name: protected-deployment
  criteria:
    - type: SpecificResource
      specificResource:
        group: apiextensions.k8s.io
        resource: customresourcedefinitions
        name: finalizerresources.apiserver.openshift.io
```

The namespace needs to be  set as "", then it works.

Reply:
Added